src/rdf4cpp/rdf/writer/Serialization.hpp

```diff
@@ -0,0 +1,178 @@
+#ifndef RDF4CPP_SERIALIZATION_HPP
+#define RDF4CPP_SERIALIZATION_HPP
+
+#include <functional>
+
+#include <rdf4cpp/rdf/Literal.hpp>
+#include <rdf4cpp/rdf/Quad.hpp>
+#include <rdf4cpp/rdf/writer/BufWriter.hpp>
+#include <rdf4cpp/rdf/writer/SerializationState.hpp>
+
+namespace rdf4cpp::rdf::writer {
+template<OutputFormat F>
+concept format_has_graph = (F == OutputFormat::NQuads || F == OutputFormat::TriG);
+
+template<OutputFormat F>
+concept format_has_prefix = (F == OutputFormat::Turtle || F == OutputFormat::TriG);
+
+struct TypeIRIPrefix {
+    std::string_view prefix;
+    std::string_view shorthand;
+};
+static constexpr std::array iri_prefixes = {
+        TypeIRIPrefix{"http://www.w3.org/2001/XMLSchema#", "xsd"},
+        TypeIRIPrefix{"http://www.w3.org/1999/02/22-rdf-syntax-ns#", "rdf"},
+};
+
+bool write_prefix_data(void *buffer, Cursor &cursor, FlushFunc flush);
+template<OutputFormat F>
+bool write_prefix(void *const buffer, Cursor &cursor, FlushFunc const flush) {
+    if constexpr (format_has_prefix<F>) {
+        return write_prefix_data(buffer, cursor, flush);
+    }
+    return true;
+}
+template<OutputFormat F>
+bool flush_state(void *const buffer, Cursor &cursor, FlushFunc const flush, SerializationState *state) {
+    if constexpr (format_has_prefix<F>) {
+        if (state == nullptr)
+            return false;
+        if (!state->active_predicate.null() || !state->active_subject.null())
+            if (!write_str(" .\n", buffer, &cursor, flush))
+                return false;
+        if (!state->active_graph.null())
+            if (!write_str("}\n", buffer, &cursor, flush))
+                return false;
+        state->active_predicate = Node::make_null();
+        state->active_subject = Node::make_null();
+        state->active_graph = Node::make_null();
+    }
+    return true;
+}
+
+template<OutputFormat F>
+bool serialize(const Quad &s, void *const buffer, Cursor &cursor, FlushFunc const flush, SerializationState *state) {
+    auto write_node = &Node::serialize_short_form;
+    if constexpr (!format_has_prefix<F>)
+        write_node = &Node::serialize;
+    auto write_predicate = [&](const Node &p) {
```

Comment:
Predicates are always IRIs.

Reply:
I noticed that there are problems further down in the code. I.e. in `Quad` it is also not `IRI` for some reason. Conflicted on whether it makes sense to fix this here. Probably leave it as is, but `assert` that the predicate is IRI instead of checking it with `if`